### DISCUSSION TITLE
FAQ/常见问题

### DISCUSSION THREAD

**Discussion by dixyes** [swow/swow#116] [General]
## cURL hook问题：使用Swow curl hook的程序崩溃，卡死

之前发现的一系列导致curl hook卡死的问题已经修复，更新到1.6.2+来修复

相关: #65 #68 #100 #233

## PIE安装失败

腾讯云的composer镜像在文档书写时存在问题，无法正产使用PIE，使用官方镜像或其他正常的镜像来使用PIE

相关: #283

## 什么时候可以正式使用？

Swow目前的代码稳定性已经实现了**基本保证**（通过[CI测试](https://github.com/swow/swow/actions)），可以保证稳定运行。

目前Swow已经进入release阶段，API/ABI稳定性根据语义版本号变更。

相关：https://github.com/swow/swow/issues/81#issuecomment-989554063 #88 #105 #120

## 如何静态编译Swow到PHP

可以使用https://github.com/crazywhalecc/static-php-cli https://github.com/dixyes/lwmbs

可以从 https://bin.dixyes.cn 下载编译好的单文件php二进制 （mainland China users only, this will be very slow for other users）

我们称呼在PHP目录树下直接同时构建PHP和Swow为in-tree构建（从Linux内核那偷的概念），in-tree构建可以将Swow静态连接到php中。
Swow的构建系统支持in-tree构建，如果执行in-tree构建失败了，请反馈bug。（TODO：写具体怎么做的文档）

相关：#119

## Swow兼容某某系统吗？

Swow对操作系统的兼容性主要受到libuv的限制，见[libuv支持平台](https://github.com/libuv/libuv/blob/v1.x/SUPPORTED_PLATFORMS.md)。
简单来说：Linux>=2.6.32 macOS>=10.15 Windows>=8 FreeBSD>=10 NDK>=r15b (Android 7.0)

相关：#107 #125

## 各种框架对Swow的支持如何？

相关信息（提醒我定期更新一下）：

hyperf: [swow-skeleton](https://github.com/hyperf/swow-skeleton)
mix-php: [api-skeleton](https://github.com/mix-php/mix/blob/master/examples/api-skeleton/README.md)
workerman: https://github.com/walkor/workerman/issues/763

相关：#62 #115

## Swow是否兼容Swoole

并不，Swow和Swoole提供了类似的设施，Swow和Swoole的PHP API不互相兼容，不能同时加载Swow和Swoole，如果你需要一个Windows下的Swoole，考虑使用[swoole-cli](https://github.com/swoole/swoole-src/releases)，[wsl](https://docs.microsoft.com/zh-cn/windows/wsl/)，[虚拟机](https://www.baidu.com/s?wd=虚拟机)，[docker desktop](https://www.docker.com/products/docker-desktop/)，[cygwin](https://www.cygwin.com/)等手段

相关：#85

## `sleep()` `msleep()` `usleep()` `Channel->pop($timeout)` 等超时提前返回

当工作负载中存在一些长CPU阻塞时，可能会发生各种睡眠/超时提前返回的现象，这是libuv的事件机制决定的，暂时没有不牺牲调度器性能的缓解方案。
如果需要高精度的延时，可以考虑使用`Watchdog`对象来强制进行事件调度，具体的使用方法参考 #160 相关讨论。
StackOverflow上的相关信息：https://stackoverflow.com/questions/42315493/why-is-libuv-timer-drifting

相关：#160

## IDE 提示

Composer包中包含了Swow扩展的stub文件，通过安装[swow/swow](https://packagist.org/packages/swow/swow)包即可为支持的IDE（PHPStorm、intephlense或者其他支持phpdoc的IDE）提供提示

相关：#164

## 为什么协程一运行就退出了

需要在底下 sleep 一段时间，或者使用其他更可靠的同步机制，详见：[Swow - 为什么协程一运行就退出了](https://docs.toast.run/swow-blog/chs/coroutine.html#%E5%8D%8F%E7%A8%8B%E7%9A%84%E8%A7%82%E6%B5%8B-%E6%8C%82%E8%B5%B7%E7%8A%B6%E6%80%81-%E4%B8%8A)

相关：#206 #169

## Vagrant或其他虚拟机共享目录下无法安装

部分虚拟机或类似设施可能通过共享目录的方式访问宿主机的文件，这种情况下由于安全考虑，一些虚拟机会禁止执行共享目录内的文件。这时，swow的composer工具可能会无法正常安装扩展。这是正常合理的安全策略限制，你可以更改你的虚拟机的相关配置或在非共享目录安装swow。

相关：#259

## 启用Swow后PHP提示PHP Warning: PHP Startup: Swow pdo_pgsql hook not enabled

如果你的Swow编译时开启了pgsql支持，在启动时环境又没有pgsql：

- 没有开启pdo扩展，提示: "pdo extension not enabled"
- 没有找到支持的libpq依赖库，提示: "libpq.so/libpq.5.dylib/lipq.dll not found" "libpq is too old"

那么启动时会提示这个，这时pdo_pgsql的hook不工作。

如果你需要支持：

- 开启PHP的[pdo扩展](https://www.php.net/manual/zh/pdo.setup.php)
- [安装libpq](https://www.postgresql.org/download/)

如果你不需要这个支持，可以配置ini：

```plain
swow.hook_pdo_pgsql = Off;
```

**Comment by dixyes** [comment]
## cURL hook: stuck, crash with Swow curl hook

A series of problems causing cURL hook stuck have been fixed since 1.6.2, update to 1.6.2+ to try again.

related: #65 #68 #100 #233

## Should I use Swow in production?

Swow now have **basic stability**, and proven to be work stably (via [continuous tests](https://github.com/swow/swow/actions)). 

Now, swow is in release, API/ABI changes will follow semantic version changes.

related：https://github.com/swow/swow/issues/81#issuecomment-989554063 #88 #105 #120

## How to statically link Swow into PHP ?

You can use https://github.com/crazywhalecc/static-php-cli https://github.com/dixyes/lwmbs

We call building an extension with PHP in PHP directory tree an "in-tree build" (this concept is stolen from kernel), in-tree build can make Swow into PHP binary.
Swow does support in-tree build, if it failed, please report the bug (TODO: write a manual for this).

related: #119

## Does Swow support some OS ?

Swow's OS requirements are based on libuv requirements, see [libuv supported platforms](https://github.com/libuv/libuv/blob/v1.x/SUPPORTED_PLATFORMS.md).
In short: Linux>=2.6.32 macOS>=10.15 Windows>=8 FreeBSD>=10 NDK>=r15b (Android 7.0)

related: #107 #125

## Common frameworks Swow support

known info (remind me to update this):

hyperf: [swow-skeleton](https://github.com/hyperf/swow-skeleton)
mix-php: [api-skeleton](https://github.com/mix-php/mix/blob/master/examples/api-skeleton/README.md)
workerman: https://github.com/walkor/workerman/issues/763

related: #62 #115

## Is Swow compatible with Swoole?

No, both Swow and Swoole provide similar facilities, their API is not compatible, Swow and Swoole cannot be loaded at the same time, if you need a PHP binary with Swoole in Windows, consider [swoole-cli](https://github.com/swoole/swoole-src/releases), [wsl](https://docs.microsoft.com/en/windows/wsl/), [virtual machines](https://www.google.com/search?q=virtual+machines), [docker desktop](https://www.docker.com/products/docker-desktop/), [cygwin](https://www.cygwin.com/), [vagrant](https://www.vagrantup.com/)

related: #85

## `sleep()` `msleep()` `usleep()` `Channel->pop($timeout)` return early

Sleep or timeout may return early with some heavy CPU workload (e.g. tons of coros rescheduling in a short period), this is due to the event scheduler mechanism of libuv. We have not found a way to work around this without harming scheduler performance at this time.
If you need high-precision sleep/timeout, you may need `Watchdog` to reschedule events, for details, see discussion in #160 
Related info on StackOverflow: https://stackoverflow.com/questions/42315493/why-is-libuv-timer-drifting

related: #160

## IDE helper

The stub file for Swow extension is included in the composer package, you can install [swow/swow](https://packagist.org/packages/swow/swow) to enable IDE Helper for supported IDE (PHPStorm, intephlense, or other IDE that is compatible with phpdoc)

related: #164

## Cannot install Swow extension under shared directory in Vagrant or other VM

Some virtual machines or similar facilities may access the host machine's files through shared directories, and due to security considerations, some virtual machines may prohibit the execution of files within shared directories. As a result, the composer tool of Swow may fail to install extension properly. This is a reasonable security policy limitation, and you can either modify the relevant configuration of your virtual machine or install Swow in a non-shared directory.

related: #259

## "PHP Warning: PHP Startup: Swow pdo_pgsql hook not enabled" after Swow enabled

If you have Swow compiled with pgsql support, but no pgsql environment when starting:

- PDO extension is enabled: "pdo extension not enabled"
- No libpq found: "libpq.so/libpq.5.dylib/lipq.dll not found" "libpq is too old"

Then you will see this warning, and pdo_pgsql hook will not work.

If you need this support:

- Enable PHP [pdo extension](https://www.php.net/manual/en/pdo.setup.php)
- [Install libpq](https://www.postgresql.org/download/)

If you don't need this support, you can add an ini configuration:

```plain
swow.hook_pdo_pgsql = Off;
```

**Comment by cexll** [comment]
今年能发布正式版吗？ 如果能大概几月出来，之前b站看了直播说今年有望发布1.0正式版十分期待，同时期待一手swoole（swow） 合体

**Comment by ikilobyte** [comment]
希望能合入php-src

**Comment by jingjingxyk** [comment]
swow 和 swoole 的关系：

PPT 第23页：
https://blog.jetbrains.com/wp-content/uploads/2022/06/20220113_jetbrains_cn_webinar_swoolenews.pdf

![image](https://github.com/swow/swow/assets/6836228/6e0fef61-fa41-4469-ab4b-a26201e688aa)
